content/blog/2019-11-08-unified-theory-gc/index.md

```diff
@@ -0,0 +1,245 @@
+
+### Split Heap and multi-Heap Collectors
+
+The most common Split-heap collector architecture is Generational collectors. Generational collectors are based on the following empirical observation that most objects are short lived, as shown in the left figure bellow. Here Y axis shows the number of bytes allocated and the X access shows the number of bytes allocated over time. ()
+
+<img src="ObjectLifetime.gif" alt="" style="width:100%"> | <img src="gen.png" alt="" style="width:100%"> 
+:-------------------------:|:-------------------------:
+[Most objects are short lived](https://www.oracle.com/webfolder/technetwork/tutorials/obe/java/gc01/index.html)  |  Generational collectors
+
+
+So Generational collectors isolated out a nursery space from the remaining mature space. Most of the time it only collects garbage from this nursery space and moves the remaining alive objects to mature space (minor collections). Once in a while it performs a garbage collection cross the whole heap to clean the mature space (major collections). An example of Generational Collectors is available in [this blog](https://blogs.msdn.microsoft.com/abhinaba/2009/03/02/back-to-basics-generational-garbage-collection/).
+
+This process can also be seen as a combination of tracing and reference counting as shown by (a) in the right figure: Reference counting is performed to track edges from mature space to nursery space. Tracing is performed within nursery space during minor collections. And finally a full tracing is performed during major collections.
+
+We can then explore different combinations of tracing and reference counting within each space. However, notice that reference counting is always used for the edges from mature to nursery space in order to avoid visiting all objects in the mature space. In fact, the authors claim that any algorithm
+that collects some subset of objects independently is fundamentally making use of reference counting. 
+
+This conclusion is further used for Multi-heap collectors, which split heaps into multiple regions and collect each region individually. They can be treated asymmetrically (as in Generational collector) or symmetrically. One of the basic extension from Generational collector is the Train algorithm. The train algorithm extended the mature space to multiple regions of different "generations", where only one region is collected at a time. This is mainly for reducing the pause time for collecting the mature space. Specifically, it divides the Mature Object Space(MOS) to cars of fixed size, which are chained to form trains. Each time, the current youngest car will be collected, until the whole first train is collected. We assume that then this process moves on to the next train, but then it is unclear to us why we need several trains instead of one train with all cars?
```

Comment:
```suggestion
This conclusion is further used for Multi-heap collectors, which split heaps into multiple regions and collect each region individually. They can be treated asymmetrically (as in Generational collector) or symmetrically. One of the basic extension from Generational collector is the Train algorithm. The train algorithm extended the mature space to multiple regions of different "generations", where only one region is collected at a time. This is mainly for reducing the pause time for collecting the mature space. Specifically, it divides the Mature Object Space (MOS) to cars of fixed size, which are chained to form trains. Each time, the current youngest car will be collected, until the whole first train is collected. We assume that then this process moves on to the next train, but then it is unclear to us why we need several trains instead of one train with all cars?
```